src/ts-ast.ts

```diff
@@ -9,117 +9,325 @@
  * rights grant found at http://polymer.github.io/PATENTS.txt
  */
 
-export interface Document {
-  kind: 'document';
-  path: string;
-  members: Array<Namespace|Class|Interface|Mixin|Function>;
+export class Document {
+  readonly kind = 'document';
+
+  constructor(
+      public path: string,
+      public members: (Namespace|Class|Interface|Mixin|Function)[] = []) {
+  }
+
+  serialize(): string {
+    return this.members.map((m) => m.serialize()).join('\n');
+  }
 }
 
-export interface Namespace {
-  kind: 'namespace';
-  name: string;
-  members: Array<Namespace|Class|Interface|Mixin|Function>;
+export class Namespace {
+  readonly kind = 'namespace';
+
+  constructor(
+      public name: string,
+      public members: (Namespace|Class|Interface|Mixin|Function)[] = []) {
+  }
+
+  serialize(depth: number = 0): string {
+    let out = ''
+    const i = indent(depth)
+    out += i
+    if (depth === 0) {
+      out += 'declare ';
+    }
+    out += `namespace ${this.name} {\n`;
+    for (const member of this.members) {
+      out += '\n' + member.serialize(depth + 1);
+    }
+    out += `${i}}\n`;
+    return out;
+  }
 }
 
-export interface Class {
-  kind: 'class';
-  name: string;
-  description: string;
-  extends: string;
-  mixins: string[];
-  properties: Property[];
-  methods: Method[];
+export class Class {
+  readonly kind = 'class';
+
+  description: string = '';
+  extends: string = '';
+  mixins: string[] = [];
+  properties: Property[] = [];
+  methods: Method[] = [];
+
+  constructor(public name: string) {
+  }
```

Comment:
whither name?

Reply:
If you annotate a constructor parameter with a visibility keyword, it is automatically assigned as a member variable.

https://www.typescriptlang.org/docs/handbook/classes.html#parameter-properties

Too obscure (second time somebody has asked about it)?